### DISCUSSION TITLE
BaseHTTPMiddleware is being called many times

### DISCUSSION THREAD

**Discussion by drnextgis** [fastapi/fastapi#9034] [Questions] [answered]
Is it expected that middleware constructor is being called as many time as many middlewares we have?

```python
import uvicorn
from fastapi import FastAPI
from starlette.middleware.base import BaseHTTPMiddleware


a = []

class DummyMiddlewareOne(BaseHTTPMiddleware):
    def __init__(self, app, p=1):
        super().__init__(app)
        a.append(p)
        print(a)

class DummyMiddlewareTwo(BaseHTTPMiddleware):
    pass

class DummyMiddlewareThree(BaseHTTPMiddleware):
    pass

app = FastAPI()
app.add_middleware(DummyMiddlewareOne)
app.add_middleware(DummyMiddlewareTwo)
app.add_middleware(DummyMiddlewareThree)

uvicorn.run(app, host='127.0.0.1', port=8000)
```

Output:

```
$ python test.py
[1]
[1, 1]
[1, 1, 1]
```

**Comment by Kludex** [comment]
Yes. It's normal. The `middleware_stack` is built every time you add a new middleware.

Reference: https://github.com/encode/starlette/blob/1222e78a15336162d190060f76c8ccee4b011d6b/starlette/applications.py#L90

**Comment by drnextgis** [comment]
Good to know, thanks! Seems very counter intuitive. I tried add logging handler inside the constructor:

```
self.logger.addHandler(logging.StreamHandler(sys.stdout))
```

and ended up having too many handlers. I believe this behavior must be documented.

**Comment by Kludex** [comment]
A discussion just started on Starlette: https://github.com/encode/starlette/discussions/1161

**Comment by thunderamur** [comment]
@drnextgis 
In your case you can use:
```python3
app = FastAPI(middleware=[
    Middleware(DummyMiddlewareOne),
    Middleware(DummyMiddlewareTwo),
])
```

**Comment by thunderamur** [comment]
It's not a problem of the FastAPI because this behavior is inherited from Starlette. I think this issue should be closed.

**Comment by ghandic** [comment]
You should be able to use `__new__` and the singleton pattern too if you're worried about multiple init

**Comment by thunderamur** [comment]
> You should be able to use `__new__` and the singleton pattern too if you're worried about multiple init

And get another problems later.
https://stackoverflow.com/questions/12755539/why-is-singleton-considered-an-anti-pattern

Do not use Singleton if you can.

**Comment by ghandic** [comment]
In this case I was suggesting to make it so you do not run initialization n times - which may be expensive. 

Essentially checking if one has already been initialized and reusing if it has.

Worth reading into the pros and cons of it and not just going off what people call "anti patterns" as it has a use, just niche.

**Comment by Kludex** [comment, marked as answer]
This behavior is not present on the latest FastAPI/Starlette. We only call the middleware once.
